Add support for local datasets. (#3)
ekumenlabs/SplaTAM: RESULTS.md

### Offline mode

Saves a directory with the RBG captured frames and the depth frames \+ a transform.json file to feed into the processing tool.  
Offline mode is currently bugged. There is an issue with the recording processing, which is not converting the depth buffer correctly into 16 bit numbers. See [this author comment](https://github.com/jc211/NeRFCapture/issues/10#issuecomment-1888164311) for more details.

## Spectacular AI

* Available for both Android and IOS.  
* IOS’ version provides significantly less configuration options than Android’s.  
  * But Android devices don’t generate depth dataset.

Recording should be taken by moving through a room while pointing the device towards the walls, and occasionally performing a panoramic view of the room to capture objects that are centered or away from the walls. The open3d capture from the ping pong table shows a good example of a trajectory the recorder should follow.

# Processing tools

## SplaTAM

### Spectacular AI dataset

In order to run SplaTAM with Spectacular AI recorded data, the first step was to process it to match the output format of NeRF Capture.

### NeRF Capture dataset

> To get the results, SplaTAM provides the [nerfcapture2dataset](https://github.com/spla-tam/SplaTAM/blob/main/scripts/nerfcapture2dataset.py) script to process the NeRF Capture output into the expected directory structure. This script works with the Online mode and should be modified to work with the Offline mode directory structure.

Reviewer: @Santoi `This script works with the Online mode and should be modified to work with the Offline mode directory structure.` what does this mean then?

Reviewer: Also, here and elsewhere, I see non-permalinks to upstream `SplaTAM`. Shouldn't these point to our fork at least? `SplaTAM` will continue changing.

Author: > what does this mean then?

I meant that for the official SplaTAM version, which still expects an online connection. The version we are working on now enables support for the Offline mode.
I'll rephrase to refer to our version then, so it also makes sense to point the links to our fork.

> After running [splatam](https://github.com/spla-tam/SplaTAM/blob/main/scripts/splatam.py) algorithm, the output is saved into a paramz.nzp format that can then be visualized with open3d viewer to get ahold of the desired point cloud, alongside a metrics plot showing the PSNR and the captured depth of each frame, and additionally, each output frame rasterized.

Reviewer: @Santoi hmm, I think it's `.npz` 
```suggestion
After running [splatam](https://github.com/spla-tam/SplaTAM/blob/main/scripts/splatam.py) algorithm, the output is saved into an paramz.npz file format that can then be visualized with open3d viewer to get ahold of the desired point cloud, alongside a metrics plot showing the PSNR and the captured depth of each frame, and additionally, each output frame rasterized.
```
Might be worth adding a link to `numpy` docs explaining the format.

Author: Done!

> # 3D Mapping Workflow (Frodo)

Reviewer: @Santoi nit: 
```suggestion
# 3D Mapping Workflow
```

Reviewer: If it helps, use an LLM to adjust the writing (punctuation, tenses, overall cohesion).

# Recording devices

> List of the recording devices used during the Frodo

Reviewer: @Santoi nit:
```suggestion
Some of the recording devices used include:
```

> * iPhone Pro  

Reviewer: @Santoi nit: links to product briefs would be nice.

  * Has lidar, which allows depth dataset recording.  
* iPhone  
  * No lidar, no depth capture.  
* Android device  
  * ARCore should suffice for the missing lidar, but there was no depth dataset saved after recording.  
  * So no depth capture.

# Recording tools

## NeRF Capture

> * Only available for IOS.  

Reviewer: @Santoi nit:
```suggestion
* Only available on iOS.  
```
Same elsewhere.

Reviewer: Also, link to app store would be nice.

* Provides 2 different recording modes:  
  * Online mode & offline mode.

### Online mode

> Requires the host to be connected into the recording device, so as to process the input while it is being captured.

Reviewer: @Santoi nit: 
```suggestion
Requires the host to be connected to the recording device, so as to process the input while it is being captured.
```

### Offline mode

Saves a directory with the RBG captured frames and the depth frames \+ a transform.json file to feed into the processing tool.  

> Offline mode is currently bugged. There is an issue with the recording processing, which is not converting the depth buffer correctly into 16 bit numbers. See [this author comment](https://github.com/jc211/NeRFCapture/issues/10#issuecomment-1888164311) for more details.

Reviewer: @Santoi nit: 
```suggestion
Offline mode is currently bugged. There is an issue with the recording process, which is not converting the depth buffer correctly into 16 bit numbers. See [this author comment](https://github.com/jc211/NeRFCapture/issues/10#issuecomment-1888164311) for more details.
```

See [NpzFile](https://numpy.org/devdocs/reference/generated/numpy.lib.npyio.NpzFile.html#numpy.lib.npyio.NpzFile) for info on `npz` format and how to load it.

### Results

First Spectacular AI results show metrics plots that look good on the first third of the processed frames, but derailed on the rest. This behavior was repeated throughout 3 captured datasets with an iPhone 14 pro.    
![image](assets/metrics.png)
This leads to the rasterized images and the point cloud visualization to look awful, since the tracking of the capturing device’s position is not correct.

![image](assets/hamilton-broken.png)

Changing the amount of frames to be processed to the point where the PSNR drops shows a much better result. It can be seen how the device tracker is within the captured environment.  
![image](assets/hamilton-02-open3d.png)
![image](assets/pool-table-open3d.png)

## Instant NGP

> \[TODO\]

Reviewer: @Santoi nit: remove?
